modules/core/src/main/java/org/apache/ignite/internal/processors/cache/warmup/LoadAllWarmUp.java

```diff
@@ -0,0 +1,207 @@
+
+            if (log.isInfoEnabled()) {
+                log.info("Start warm-up cache group, starting from index [name=" + grp.cacheOrGroupName()
+                    + ", partCnt=" + partCnt + ", pageCnt=" + pageCnt + ']');
+            }
+
+            PageMemoryEx pageMemEx = (PageMemoryEx)region.pageMemory();
+
+            long grpPageCnt = 0;
+
+            // Starting with index.
+            for (int i = -1; i < partCnt - 1; i++, allPartCnt++) {
+                long pageId = pageMemEx.partitionMetaPageId(grp.groupId(), (i == -1 ? INDEX_PARTITION : i));
+                long pages = grp.shared().pageStore().pages(grp.groupId(), (i == -1 ? INDEX_PARTITION : i));
+
+                for (long j = 0; j < pages && grpPageCnt < pageCnt; j++, pageId++, grpPageCnt++, allPageCnt++) {
+                    if (stop) {
+                        if (log.isInfoEnabled()) {
+                            log.info("Stop warm-up cache group with loading statistics [name="
+                                + grp.cacheOrGroupName() + ", partCnt=" + (i + 1) + "partPageCnt=" + j +
+                                ", grpPageCnt=" + grpPageCnt + ", allPartCnt=" + allPartCnt + ", allPageCnt=" +
+                                allPageCnt + ']');
+                        }
+
+                        return;
+                    }
+
+                    long pagePtr = -1;
+
+                    try {
+                        pagePtr = pageMemEx.acquirePage(grp.groupId(), pageId);
+                    }
+                    finally {
+                        if (pagePtr != -1)
+                            pageMemEx.releasePage(grp.groupId(), pageId, pagePtr);
+                    }
+                }
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void stop() throws IgniteCheckedException {
+        stop = true;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        return S.toString(LoadAllWarmUp.class, this);
+    }
+
+    /**
+     * Getting count of pages available for loading into data region.
+     *
+     * @param region Data region.
+     * @return Count(non-negative) of pages available for loading into data region.
+     */
+    protected long availableLoadPageCount(DataRegion region) {
+        long maxSize = region.config().getMaxSize();
+        long curSize = region.pageMemory().loadedPages() * region.pageMemory().systemPageSize();
+
+        return Math.max(0, (maxSize - curSize) / region.pageMemory().systemPageSize());
+    }
+
+    /**
+     * Calculation of cache groups, count of partitions and pages that can load
+     * into data region. Calculation starts and includes an index partition for
+     * each group.
+     *
+     * @param region Data region.
+     * @return Mapping: {cacheGrp -> {partCnt, pageCnt}}.
+     * @throws IgniteCheckedException – if faild.
+     */
+    protected Map<CacheGroupContext, T2<Integer, Long>> loadDataInfo(
```

Comment:
Maybe is it better instead of this unspecified map to have some list of specific class like WarmUpPartitionInfo(cacheGroupContext, partitionId, MaxPagesToLoad)?